cmd/thanos/receive.go

```diff
@@ -251,34 +296,46 @@ func runReceive(
 	level.Debug(logger).Log("msg", "setting up grpc server")
 	{
 		var (
-			s   *grpc.Server
-			l   net.Listener
-			err error
+			s *grpc.Server
+			l net.Listener
 		)
+		startGRPC := make(chan struct{})
 		g.Add(func() error {
-			<-dbOpen
-
-			l, err = net.Listen("tcp", grpcBindAddr)
-			if err != nil {
-				return errors.Wrap(err, "listen API address")
-			}
-
-			db := localStorage.Get()
-			tsdbStore := store.NewTSDBStore(log.With(logger, "component", "thanos-tsdb-store"), reg, db, component.Receive, lset)
-
 			opts, err := defaultGRPCServerOpts(logger, cert, key, clientCA)
 			if err != nil {
 				return errors.Wrap(err, "setup gRPC server")
 			}
-			s := newStoreGRPCServer(logger, reg, tracer, tsdbStore, opts)
 
-			level.Info(logger).Log("msg", "listening for StoreAPI gRPC", "address", grpcBindAddr)
-			return errors.Wrap(s.Serve(l), "serve gRPC")
+			for range dbReady {
+				if s != nil {
+					s.Stop()
+				}
+				l, err = net.Listen("tcp", grpcBindAddr)
+				if err != nil {
+					return errors.Wrap(err, "listen API address")
+				}
+				tsdbStore := store.NewTSDBStore(log.With(logger, "component", "thanos-tsdb-store"), nil, localStorage.Get(), component.Receive, lset)
+				s = newStoreGRPCServer(logger, &unRegisterer{reg}, tracer, tsdbStore, opts)
+				startGRPC <- struct{}{}
+			}
+			return nil
 		}, func(error) {
 			if s != nil {
 				s.Stop()
 			}
+			close(startGRPC)
 		})
+		// We need to be able to start and stop the gRPC server
```

Comment:
I don't understand why we need this to be separate from the above run group. It seems the above one immediately triggers this one when sending to the `startGRPC` channel. What would the difference be if we just combined the two?

Reply:
the gRPC server needs to be restartable, so this means that the routine that stops and starts the server should not block but `s.Serve(l)` is blocking, so we want to run the gRPC server in it's own go-routine. "combining the two" means that rather than use a run group we simply do `go func() {if err := s.Serve(l)...}()`. Either way there's a new go routine. The difference is that now the gRPC server can't cause the run group to exit when it experiences an error.

Reply:
I don't see why it needs to be restarted. I'm probably totally missing it :slightly_smiling_face: .

Reply:
We need to restart because the actual tsdb instance used by in `tsdbStore := store.NewTSDBStore(log.With(logger, "component", "thanos-tsdb-store"), nil, localStorage.Get(), component.Receive, lset)` is rotated after ever re-sharding. So before every restart we create a new `tsdbStore`, make a new gRPC server from that `tsdbStore` instance and start the server

Reply:
got it, that makes perfect sense